src/python/pants/option/options_fingerprinter.py

```diff
@@ -16,7 +16,24 @@
                                        target_option)
 
 
+# TODO: this class's methods could be wrapped in @memoized_method -- we would have to manage the
+# cache size somehow if this is done or it might blow up.
 class Encoder(json.JSONEncoder):
+  def _encode_dict_wrapping_keys(self, d):
+    return {self.encode(k):v for k, v in d.items()}
+
+  def encode(self, o):
+    if isinstance(o, (tuple, set)):
+      o = list(o)
+    elif isinstance(o, dict):
+      o = self._encode_dict_wrapping_keys(o)
+    elif len(o) == 1:
+      only_element = next(iter(o))
+      if isinstance(only_element, dict):
+        o = [self._encode_dict_wrapping_keys(only_element)]
+
+    return super(Encoder, self).encode(o)
+
   def default(self, o):
     if o is UnsetBool:
```

Comment:
Why aren't you using `default`? It seems to me your conditionals should move down here since this is the documented way to hook things.

Reply:
modifying `default` resulted in the same error for me until I tried overriding `encode()` -- that's not saying it's the right way whatsoever. I will look into why I was seeing that behavior and whether this change can be done entirely in `default`.

Reply:
This is green against your tests mod sha changes for the `{'a': 3}` and `[{'a': 3}]` cases:
```diff
diff --git a/src/python/pants/option/options_fingerprinter.py b/src/python/pants/option/options_fingerprinter.py
index 224921056..f8ce100db 100644
--- a/src/python/pants/option/options_fingerprinter.py
+++ b/src/python/pants/option/options_fingerprinter.py
@@ -10,6 +10,8 @@ import os
 from builtins import object, open, str
 from hashlib import sha1
 
+from future.moves import collections
+
 from pants.base.build_environment import get_buildroot
 from pants.base.hash_utils import stable_json_hash
 from pants.option.custom_types import (UnsetBool, dict_with_files_option, dir_option, file_option,
@@ -20,6 +22,8 @@ class Encoder(json.JSONEncoder):
   def default(self, o):
     if o is UnsetBool:
       return '_UNSET_BOOL_ENCODING'
+    elif isinstance(o, collections.Iterable) and not isinstance(o, collections.Mapping):
+      return list(o)
     return super(Encoder, self).default(o)
```

Reply:
Simplify!

Reply:
I think the reason it wasn't working was because I hadn't figured out that `fingerprint` wraps everything into a list yet, so I kept getting the same dict error and didn't at that point put any print statements to show it was working. Also, other code online used `encode()`. Thanks for figuring this out.